Sources/Sentry/SentryProfiler.mm

```diff
@@ -306,126 +349,97 @@ + (void)startWithHub:(SentryHub *)hub
-#    endif // SENTRY_HAS_UIKIT
-
-    [_gCurrentProfiler start];
-
-    _gCurrentProfiler->_timeoutTimer = [SentryDependencyContainer.sharedInstance.timerFactory
-        scheduledTimerWithTimeInterval:kSentryProfilerTimeoutInterval
-                                target:self
-                              selector:@selector(timeoutAbort)
-                              userInfo:nil
-                               repeats:NO];
-#    if SENTRY_HAS_UIKIT
-    [[NSNotificationCenter defaultCenter] addObserver:self
-                                             selector:@selector(backgroundAbort)
-                                                 name:UIApplicationWillResignActiveNotification
-                                               object:nil];
-#    endif // SENTRY_HAS_UIKIT
-}
-
-+ (void)stop
-{
-    std::lock_guard<std::mutex> l(_gProfilerLock);
-
-    if (!_gCurrentProfiler) {
-        SENTRY_LOG_WARN(@"No current global profiler manager to stop.");
-        return;
-    }
-    if (![_gCurrentProfiler isRunning]) {
-        SENTRY_LOG_WARN(@"Current profiler is not running.");
-        return;
-    }
-
-    [self stopProfilerForReason:SentryProfilerTruncationReasonNormal];
+    trackProfilerForTracer(_gCurrentProfiler, tracer);
 }
 
-+ (BOOL)isRunning
++ (BOOL)isCurrentlyProfiling
 {
     std::lock_guard<std::mutex> l(_gProfilerLock);
     return [_gCurrentProfiler isRunning];
 }
 
-+ (SentryEnvelopeItem *)createProfilingEnvelopeItemForTransaction:(SentryTransaction *)transaction
++ (nullable SentryEnvelopeItem *)createProfilingEnvelopeItemForTransaction:
+    (SentryTransaction *)transaction
 {
-    const auto profileID = [[SentryId alloc] init];
-    const auto payload = [self serializeForTransaction:transaction profileID:profileID];
+    const auto profiler = profilerForFinishedTracer(transaction.trace);
+    if (!profiler) {
+        SENTRY_LOG_WARN(@"Expected a profiler for tracer id %@ but none was found",
+            transaction.trace.traceId.sentryIdString);
+        return nil;
+    }
+
+    const auto payload = [profiler serializeForTransaction:transaction];
 
 #    if defined(TEST) || defined(TESTCI)
     [NSNotificationCenter.defaultCenter postNotificationName:@"SentryProfileCompleteNotification"
                                                       object:nil
                                                     userInfo:payload];
 #    endif // defined(TEST) || defined(TESTCI)
 
-    return [self envelopeItemForProfileData:payload profileID:profileID];
+    const auto JSONData = [SentrySerialization dataWithJSONObject:payload];
+    if (JSONData == nil) {
+        SENTRY_LOG_DEBUG(@"Failed to encode profile to JSON.");
+        return nil;
+    }
+
+    const auto header = [[SentryEnvelopeItemHeader alloc] initWithType:SentryEnvelopeItemTypeProfile
+                                                                length:JSONData.length];
+    return [[SentryEnvelopeItem alloc] initWithHeader:header data:JSONData];
```

Comment:
This was the only place this method was called, so I just inlined the implementation here.